### PR TITLE
Add equality comparison operators to Query and underlying required objects

### PR DESCRIPTION
**Pull request by gbiggs**
### Equality comparison operators for Query

This PR adds the equal and not-equal operators to the schedule `Query` object.
It also adds them to all objects used internally (recursive) by `Query`.
This allows `Query`s to be directly compared for equality, which supports operations such as identifying if two schedule mirrors are requesting the same query.

### Implementation description

In general, the operators are implemented by member-wise comparisons of each object's members.
Shapes are a special case.
A template function and a "register" function are used to reduce the amount of code that needs to be added for each new shape type.

All not-equals operators are implemented as the inverse of the equality operators to reduce the code that must be maintained.

### How to test

Execute the unit tests with the `[query]` tag to filter down to just the query tests.

### CODE REVIEW

**Review by mxgrey — CHANGES_REQUESTED** [review]
This is generally looking really good!

I have a few nitpick requests, and then there's one actually important request regarding the `Timespan` comparison. I've pushed a unit test that shows the `Timespan` comparison currently failing for an edge case.

I've also merged in the latest `main` which introduces code linting, and some of the changes in this PR are violating the style guide, particularly for line length.

File: rmf_traffic/include/rmf_traffic/geometry/Shape.hpp [hunk 1]
```diff
@@ -108,6 +115,20 @@ FinalShapePtr make_final(const T& shape)
   return std::make_shared<FinalShape>(shape.finalize());
 }
 
+//==============================================================================
+template<typename T>
+std::function<bool(const Shape& other)> make_equality_comparitor(const T& myself)
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic/include/rmf_traffic/geometry/Shape.hpp]: Fixed in d7de03f7dd3880b4bfbfb84f2221c4c0cee21e56.

File: rmf_traffic/test/unit/schedule/test_Query.cpp [hunk 2]
```diff
@@ -61,18 +61,158 @@ SCENARIO("Test Query API")
...
+    auto now_plus_1min = now + 1min;
+    // TODO(Geoff): These tests are not exhaustive. Make them so.
+    WHEN("Both queries are empty")
+    {
+      rmf_traffic::schedule::Query query1 =
+        rmf_traffic::schedule::make_query({});
+      rmf_traffic::schedule::Query query2 =
+        rmf_traffic::schedule::make_query({});
+
+      REQUIRE(query1 == query2);
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic/test/unit/schedule/test_Query.cpp:86]: Done in 26a7cf0.

File: rmf_traffic/include/rmf_traffic/geometry/Shape.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic/include/rmf_traffic/geometry/Shape.hpp]: Done in d7de03f7dd3880b4bfbfb84f2221c4c0cee21e56.

File: rmf_traffic/src/rmf_traffic/schedule/Query.cpp [hunk 3]
```diff
@@ -451,6 +484,40 @@ auto Query::Spacetime::timespan() const -> const Timespan*
   return nullptr;
 }
 
+//==============================================================================
+bool operator==(
+  const Query::Spacetime::Timespan& lhs,
+  const Query::Spacetime::Timespan& rhs)
+{
+  return lhs.maps() == rhs.maps() &&
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic/src/rmf_traffic/schedule/Query.cpp]: Done in f13acda.

**Review by mxgrey — APPROVED** [review]
This looks good! It seems there are just some style complaints from uncrustify to fix.
